contributors/design-proposals/storage/dynamic_volume_limit.md

```diff
@@ -0,0 +1,102 @@
+# Dynamic attached volume limits
+
+## Goals
+
+Currently the number of volumes attachable to a node is either hardcoded or only configurable via an environment variable. Also
+existing limits only apply to well known volume types like EBS, GCE and is not available to all volume plugins.
+
+This proposal enables any volume plugin to specify those limits and also allows same volume type to have different volume
+limits depending on type of node.
+
+## Implementation Design
+
+### Prerequisite
+
+* This feature will be protected by an alpha feature gate, so as API and CLI changes needed for it. We are planning to call
+  the feature `DynamicVolumeThreshold`.
+
+### API Changes
+
+There is no API change needed for this feature. However existing `node.Status.Capacity` will
+be extended to cover volume limits available on the node too.
+
+The key name that will store volume will be start with prefix `storage-`. So for example:
+
+* `storage-aws-ebs`
+* `storage-gce-pd`
+* `storage-csi-<driver_name>`
+
+`IsScalarResourceName` check will be extended to cover storage limits:
+
+```go
+func IsStorageResourceName(name v1.ResourceName) bool {
+    return strings.HasPrefix(string(name), v1.ResourceStoragePrefix)
+}
+
+// Extended and Hugepages resources
+func IsScalarResourceName(name v1.ResourceName) bool {
+    return IsExtendedResourceName(name) || IsHugePageResourceName(name) ||
+        IsPrefixedNativeResource(name) || IsStorageResourceName(name)
+}
+```
+
+#### Alternative to using "storage-" prefix
+We also considered using currently defined `GetPluginName` interface(of Volume Plugins) for using as key in the `node.Status.Capacity`. Ultimately
+we decided against using it, because most in-tree plugins start with `kubernetes.io/` and we needed a uniform way to identify storage
+related capacity data in `node.Status`.
+
+### Changes to CSI
+
+Corresponding change is being made to CSI so as CSI drivers can output
+volume limits as part of Node capabilities - https://github.com/container-storage-interface/spec/pull/221
+
+csi node-registrar will be updated to set the `node.Status` property after fetching
+volume limit from CSI driver.
+
+### Changes to scheduler
+
+Scheduler will retrieve available attachable limit on a node from `node.Status.Capacity` and store it in `nodeInfo` cache. Volume
+limits will be treated like any other scalar resource.
+
+Existing `NewMaxPDVolumeCountPredicate` will be extended to calculate volumes in-use on a node by counting all scheduled pods on the node
+and then comparing it with limit available from `nodeInfo`.
+
+To be backward compatible - the scheduler will fallback to older logic, if no limit is set in `node.Status.Capacity` for AWS, GCE and Azure volume
```

Comment:
i think it would be better to use node.Status.Allocatable more than node.Status.Capacity.